module/builder/consensus/builder.go

```diff
@@ -445,20 +464,35 @@ func (b *Builder) getInsertableReceipts(parentID flow.Identifier) ([]*flow.Execu
 		return nil, fmt.Errorf("failed to retrieve reachable receipts from memool: %w", err)
 	}
 
+	filteredReceipts := make([]*flow.ExecutionReceiptMeta, 0, len(receipts))
+	results := make([]*flow.ExecutionResult, 0)
+	for _, receipt := range receipts {
+		resultID := receipt.ExecutionResult.ID()
+		if _, inserted := includedResults[resultID]; !inserted {
+			results = append(results, &receipt.ExecutionResult)
+			includedResults[resultID] = struct{}{}
+		}
+
+		filteredReceipts = append(filteredReceipts, receipt.Meta())
+	}
+
 	// don't collect more than maxReceiptCount receipts
-	if uint(len(receipts)) > b.cfg.maxReceiptCount {
-		receipts = receipts[:b.cfg.maxReceiptCount]
+	if uint(len(filteredReceipts)) > b.cfg.maxReceiptCount {
+		filteredReceipts = filteredReceipts[:b.cfg.maxReceiptCount]
 	}
```

Comment:
⚠️  this might drop some receipts whose result has been added to the `results`. 

I would suggest to combine [line467 to line 487](https://github.com/onflow/flow-go/pull/462/files#diff-b8e03abbeb857725873745cd0ef3acccfda614eb3ba07c801f2072239ac7eb0eR467-R481) into a single function:

```
insertables := findInsertables(includedResults, receipts, b.cfg.maxReceiptCount)
return insertables
```

Easier to write tests to verify